contracts/strategy/ConverterStrategyBase.sol

```diff
@@ -0,0 +1,332 @@
+    _doHardWork(false);
+    if (amount == 0) return;
+
+    address _asset = asset;
+    uint assetBalanceBefore = _balance(_asset);
+
+    address[] memory tokens = _depositorPoolAssets();
+    uint len = tokens.length;
+    uint[] memory tokenAmounts = new uint[](len);
+    uint assetAmountForToken = amount / len;
+
+    for (uint i = 0; i < len; ++i) {
+      address token = tokens[i];
+      if (token == _asset) {
+        tokenAmounts[i] = assetAmountForToken;
+      } else {
+        uint tokenBalance = _balance(token);
+
+        if (tokenBalance >= assetAmountForToken) { // we already have enough tokens
+          tokenAmounts[i] = tokenBalance;
+
+        } else { // we do not have enough tokens - borrow
+         _openPosition(_asset, assetAmountForToken - tokenBalance, token, ITetuConverter.ConversionMode.AUTO_0);
+          tokenAmounts[i] = _balance(token);
+        }
+      }
+    }
+
+    _depositorEnter(tokenAmounts);
+    _investedAssets += (assetBalanceBefore - _balance(_asset));
+  }
+
+  /// @dev Withdraw given amount from the pool.
+  function _withdrawFromPoolUniversal(uint amount, bool emergency) internal {
+    if (amount == 0) return;
+
+    if (!emergency) _doHardWork(false);
+
+    address _asset = asset;
+    uint assetBalanceBefore = _balance(_asset);
+
+    if (emergency) {
+      _depositorEmergencyExit();
+    } else {
+      uint liquidityAmount = amount * _depositorLiquidity() / _investedAssets;
+      _depositorExit(liquidityAmount);
+    }
+
+    address[] memory tokens = _depositorPoolAssets();
+    uint len = tokens.length;
+
+    for (uint i = 0; i < len; ++i) {
+      address borrowedToken = tokens[i];
+      if (_asset != borrowedToken) {
+        _closePosition(_asset, borrowedToken, _balance(borrowedToken));
+      }
+    }
+
+    uint amountReceived = _balance(_asset) - assetBalanceBefore;
+    _investedAssets -= amountReceived;
+
+  }
+
+  /// @dev Withdraw given amount from the pool.
+  function _withdrawFromPool(uint amount) override internal virtual {
+    _withdrawFromPoolUniversal(amount, false);
+  }
+
+  /// @dev Withdraw all from the pool.
+  function _withdrawAllFromPool() override internal virtual {
+    _withdrawFromPoolUniversal(_investedAssets, false);
+  }
+
+  /// @dev If pool support emergency withdraw need to call it for emergencyExit()
+  function _emergencyExitFromPool() override internal virtual {
+    _withdrawFromPoolUniversal(_investedAssets, true);
+  }
+
+
+  function _recycle(address[] memory tokens, uint[] memory amounts) internal {
+    require(tokens.length == amounts.length, "SB: Arrays mismatch");
+    uint len = tokens.length;
+
+    for (uint i = 0; i < len; ++i) {
+      address token = tokens[i];
+      uint amount = amounts[i];
+
+      if (amount > 0) {
+        uint amountToCompound = amount * compoundRatio / COMPOUND_DENOMINATOR;
+        if (amountToCompound > 0) {
+          tetuLiquidator.liquidate(token, asset, amountToCompound, LIQUIDATION_SLIPPAGE);
```

Comment:
tetuLiquidator to local var

Reply:
and asset

Reply:
1) where approve?
2) liquidation more complex process - need to check routes, make properly decision need to liquidate or not, assume that some dust can exist

Reply:
check Forwarder contract how it should be handled